mesonbuild/interpreter/interpreter.py

```diff
@@ -2079,11 +2080,26 @@ def func_install_headers(self, node: mparser.BaseNode,
             if os.path.isabs(install_subdir):
                 mlog.deprecation('Subdir keyword must not be an absolute path. This will be a hard error in the next release.')
 
-        h = build.Headers(source_files, install_subdir, kwargs['install_dir'],
-                          kwargs['install_mode'], self.subproject)
-        self.build.headers.append(h)
+        if kwargs['preserve_path']:
+            dirs = collections.defaultdict(list)
+            ret_headers = []
+            for file in source_files:
+                dirname = os.path.dirname(file.fname)
+                dirs[dirname].append(file)
 
-        return h
+            for childdir in dirs:
+                h = build.Headers(dirs[childdir], os.path.join(install_subdir, childdir), kwargs['install_dir'],
+                                  kwargs['install_mode'], self.subproject)
+                ret_headers.append(h)
+                self.build.headers.append(h)
+
+            return ret_headers
+        else:
+            h = build.Headers(source_files, install_subdir, kwargs['install_dir'],
```

Comment:
Since this code is essentially the same as the other branch, what are your thoughts on making `dirs` without a preserce_path kwarg be `{'': source_files}`?

Reply:
Hmm, does it pass CI if you leave the dot out?

Reply:
yes, seems to windows related. didn't occur on linux